### DISCUSSION TITLE
How can I run a cronjob inside a useeffect only from the OnClick command?

### DISCUSSION THREAD

**Discussion by Elindo586** [vercel/next.js#51175] [Help] [answered]
### Summary

So the idea is to run the functions inside the useEffect only when I do the Onclick button.

From the code below it has a couple problems.

1.`  console.log({ current1 });`  inside the useEffect isn't showing up.

2. `cronJob1.start();`  inside the useEffect seems to run as soon as the page loads..

The error I am getting is creation of duplicate content, which leads me to believe the problem is the `cronJob1.start()`  going by itself right away... which maybe something wrong with the UseEffect?

```
import React from "react";
import { useEffect, useState } from "react";

import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";

import { CronJob } from "cron";

import db from "../../utils/text.json";

export async function getStaticProps() {
  return {
    props: { db },
  };
}

export default function TweetTest({ db }) {
  const [current1, setCurrent1] = useState(0);

  useEffect(() => {
    console.log({ current1 });

    const getText = () => {
      const target = db.find((item) => item.id === "1");
      if (!target) return "";
      return target.text;
    };
    const cronJob1 = new CronJob("17 10 * * * ", async () => {
      try {
        await fetch("/api/twitterClient", {
          method: "POST",
          headers: {
            "content-type": "application/json",
          },
          body: JSON.stringify({ message: getText() }),
        });
      } catch (e) {
        console.log(e); // fetch doesn't really throw on 4xx range
      }
    });
    cronJob1.start();
  }, [current1, db]);

  return (
    <div className="main-about">
      <Container>
        <div> {} </div> <br />
        <div>
          {/* Hello there {current1} */}
          <button
            type="button"
            className="btn btn-primary"
            onClick={() => setCurrent1(1)}
          >
            Click
          </button>
        </div>
      </Container>
    </div>
  );
}

```

### Example

https://github.com/Elindo586/nextprojectd/blob/main/pages/twitter/test.js

**Comment by Elindo586** [comment]
I moved the `console.log({ current1 });`  into its own useEffect and kind of works even when it logs twice in the console.

The rest of items seem like they need to stay in the same useEffect because of dependencies ..

It seems like it keep firing the useEffect on load even when I am using dependencies in the brackets.

**Comment by Elindo586** [comment]
Ha... it worked...

changed the current value to blank... so it only runs when the value is set to something....

**Comment by Elindo586** [comment]
No... it's not working.... still firing the cronjob in useEffect on load.

**Comment by icyJoseph** [comment, marked as answer]
`useEffect` is to sync with the outside world. It runs after the component mounts, and then every time its dependencies change.

In DevMode theres something called StrictMode which makes it run twice as much too, to help you identify potential errors.

I'd say either, handle it carefully in the effect, or just trigger the work on click.

I am not sure about using that cron package in the browser, but I think you wanna do:

```js
export default function TweetTest({ db }) {
  // maybe start current1 as null
  const [current1, setCurrent1] = useState(null);

  useEffect(() => {
    console.log({ current1 });

    // terminate if there's nothing on current 1
    if (current1 === null) return;

    const getText = () => {
      const target = db.find((item) => item.id === "1");
      if (!target) return "";
      return target.text;
    };

    const controller = new AbortController();
    const cronJob1 = new CronJob("06 10 * * * ", async () => {
      try {
        await fetch("/api/twitterClient", {
          method: "POST",
          headers: {
            "content-type": "application/json",
          },
          body: JSON.stringify({ message: getText() }),
          signal: controller.signal,
        });
      } catch (e) {
        if (controller.signal.aborted) return; // do nothing
        console.log(e); // fetch doesn't really throw on 4xx range
      }
    });

    cronJob1.start();

    return () => {
      cronJob1.stop();
      controller.abort();
    };
  }, [current1, db]);

  return (
    <div className="main-about">
      <Container>
        <div> {} </div> <br />
        <div>
          {/* Hello there {current1 ?? ''} */}
          <button
            type="button"
            className="btn btn-primary"
            onClick={() => setCurrent1(1)}
          >
            Click
          </button>
        </div>
      </Container>
    </div>
  );
}
```

I'd recommend reading up on Abort Controllers, and clean up functions on use-effect: https://react.dev/reference/react/useEffect#connecting-to-an-external-system

**Reply by Elindo586** [reply]
Seems like a never ending learning..

May I ask how you manage to get the cronjob to start onclick?   The dependencies were included in the original code `[current1, db])`  , but cron still ran inside UseEffect...

I suppose this is the code to ensure it would stop:
```
return () => {
      cronJob1.stop();
      controller.abort();
    };
```

**Reply by icyJoseph** [reply]
I think I did not understand correctly the feature here, so please dismiss the click comment. And yes that's the code that runs a clean up and stops the cronjob, and any on-going fetch request.

**Reply by gerardnico** [reply]
Is this really a cron job? Taking a look at the code, it seems to have no job table and used [setTimeout](https://datacadamia.com/web/browser/settimeout)?

[useEffect](https://react.dev/reference/react/useEffect) runs:
* its body: after the tree rendering on mount, and when the dependencies changes
* the returned function as cleanup, on unmount.

In this case, it means that if you navigate away from your component, you stop the job.
Not really a good place to run a cronjob.

**Reply by icyJoseph** [reply]
Yeah, I googled that package and it looks like a Nodejs kind of thing, that happens to support browsers with timers. That's a yellow flag at best.

Now with memory saver features, the page might be killed arbitrarily by Chrome.

And there's also drift and how timers are handled when in the background...

And even if run on Node, with Vercel you have to use their Cron jobs API instead... https://github.com/vercel/examples/tree/main/solutions/cron repo example

I didn't want to go into details of the design solution, given the context, but yeah, you are correct.

**Comment by Elindo586** [comment]
Well... look at that...

It doesn't work from my localhost if I shutdown the browser...

and it doesn't work at Vercel at all...

Guess I am gonna have to look into the Vercel's cron docs... but  I don't have much time left this week....
